### PR TITLE
feat: accept OAS 3.1 documents

### PR DESCRIPTION
**Pull request by P0lip**
Addresses https://github.com/stoplightio/platform-internal/issues/5884
Related to  https://github.com/stoplightio/prism/issues/1691

If you're a PF member, I suggest to read https://github.com/stoplightio/platform-internal/issues/5879 first as this issue sheds some light on the whole OAS 3.1 approach we have at Stoplight. Do note that I eventually decided to use Draft 7 as the default.

List of OAS 3.1 related changes in bumped Stoplight-owned packages

- `@stoplight/types` https://github.com/stoplightio/types/compare/9723309334ceb84b1c895844ae750318348682c2~1...a7230501b1799dc90417653b04205696bebb3812
- `@stoplight/json-schema-sampler` https://github.com/stoplightio/json-schema-sampler/compare/df1bf2be0ec5646bb0b7c886145259baa443358a~1...c88c37cdea8f65549c5e1e0c4fe8500cfdefdf11 + there's still one [PR](https://github.com/stoplightio/json-schema-sampler/pull/12) open but it's not blocking this one 
- `@stoplight/http-spec` https://github.com/stoplightio/http-spec/commit/faa93c8ce1ac692a15c97e468010d520af46e132 https://github.com/stoplightio/http-spec/compare/553f44751292840a4953b2c53ec991515b186cde~1...eb6864510f1077660db37f74a8f7b553785d468e

in short:
- OAS Schema Objects -> JSON Schema Draft 7 or higher
- `@stoplight/json-schema-sampler` replaced `openapi-sampler`
- Ajv upgraded to v8
- a few unrelated bugfixes, most of them are annotated with comments

### CODE REVIEW

File: test-harness/specs/proxy/proxy.validation.httpHeader.txt [hunk 1]
```diff
@@ -31,6 +31,6 @@ proxy -p 4010 ${document} http://httpbin.org
 curl -i http://localhost:4010/json
 ====expect====
 HTTP/1.1 200 OK
-sl-violations: [{"location":["response","body"],"severity":"Error","code":"required","message":"should have required property 'title'"},{"location":["response","body"],"severity":"Error","code":"required","message":"should have required property 'description'"}]
+sl-violations: [{"location":["response","body"],"severity":"Error","code":"required","message":"must have required property 'title'"},{"location":["response","body"],"severity":"Error","code":"required","message":"must have required property 'description'"}]
 
-{"slideshow":{"author":"Yours Truly","date":"date of publication","slides":[{"title":"Wake up to WonderWidgets!","type":"all"},{"items":["Why <em>WonderWidgets</em> are great","Who <em>buys</em> WonderWidgets"],"title":"Overview","type":"all"}],"title":"Sample Slide Show"},"priority":0}
```
Comment (P0lip) [inline test-harness/specs/proxy/proxy.validation.httpHeader.txt:36]: This output was actually invalid.
This `priority` property was included by `ajv` at the time we validated the input here https://github.com/stoplightio/prism/blob/master/packages/http/src/validator/validators/utils.ts#L46

File: packages/http/src/mocker/__tests__/HttpMocker.spec.ts [hunk 2]
```diff
@@ -527,8 +527,8 @@ describe('mocker', () => {
                 },
               });
 
-              it('should be set to null', () =>
-                assertRight(eitherResponse, response => expect(response.body).toHaveProperty('age', null)));
+              it('should be set to number', () =>
```
Comment (P0lip) [inline packages/http/src/mocker/__tests__/HttpMocker.spec.ts:530]: `@stoplight/json-schema-sampler` supports arrayish type unlike openapi-sampler

File: test-harness/specs/references/E2E-Multiple-Refs-Handling.oas3.txt [hunk 3]
```diff
@@ -32,4 +32,4 @@ curl -i http://localhost:4010/
 ====expect====
 HTTP/1.1 200 OK
 
-{"first":{"output":null},"second":{"output":null}}
+{"first":{"output":"string"},"second":{"output":"string"}}
```
Comment (P0lip) [inline test-harness/specs/references/E2E-Multiple-Refs-Handling.oas3.txt:35]: `@stoplight/json-schema-sampler` supports arrayish type unlike openapi-sampler

File: packages/http/src/validator/validators/utils.ts [hunk 4]
```diff
@@ -5,16 +5,14 @@ import { pipe } from 'fp-ts/function';
...
-  schemaId: 'auto',
-  coerceTypes: false,
-});
+const ajv = new Ajv({ allErrors: true, messages: true, coerceTypes: true, strict: false });
+const ajvNoCoerce = new Ajv({ allErrors: true, messages: true, coerceTypes: false, strict: false });
+
+addFormats(ajv);
+addFormats(ajvNoCoerce);
 
 export const convertAjvErrors = (errors: NonEmptyArray<ErrorObject>, severity: DiagnosticSeverity, prefix?: string) =>
```
Comment (P0lip) [inline packages/http/src/validator/validators/utils.ts:17]: We could potentially use https://github.com/stoplightio/better-ajv-errors

**Review by philsturgeon — APPROVED** [review]
LGTM. 

This might end up needing to be a v5.x if we consider the change in error messages "breaking" but tbh I'd be fine with a v4.2.0 as its pretty minor. 

Do you think we should beta or is it safe enough to just go for it?

**Comment by chohmann** [comment]
IMO changing message doesn't seem like a breaking change.

As far as a beta version I think that's really up to you guys since it's your project.

**Comment by philsturgeon** [comment]
We could probably skip beta if we’re already using http-spec elsewhere in production. I feel like that’s been doing most of the heavy lifting.

**Comment by P0lip** [comment]
> We could probably skip beta if we’re already using http-spec elsewhere in production.

We are not yet. Prism is going to be the first project using it.
Then next project using the newer version of http-spec is going to be Spectral (the converter piece from http-spec for ajv v8), and lastly Studio.

@philsturgeon do you want to play around with it before I merge it or are we good to go?

**Comment by philsturgeon** [comment]
Let’s just merge it. I asked around and nobody has OAS 3.1 because nobody has released any tools that work with it yet. Let’s fix that. 

--
Phil Sturgeon
Product @ Stoplight.io
